x/delegation/keeper/delegation_state.go

```diff
@@ -11,97 +11,133 @@ import (
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
-// UpdateStakerDelegationTotalAmount function is used to update the delegation total amount of the specified staker and asset.
-// The input `opAmount` represents the values that you want to add or decrease, using positive or negative values for increasing and decreasing, respectively.
-// The function will calculate and update new state after a successful check.
-// The function will be called when there is delegation or undelegation related to the specified staker and asset.
-func (k *Keeper) UpdateStakerDelegationTotalAmount(ctx sdk.Context, stakerID string, assetID string, opAmount sdkmath.Int) error {
-	if opAmount.IsNil() || opAmount.IsZero() {
-		return nil
-	}
-	// use stakerID+'/'+assetID as the key of total delegation amount
+type DelegationOpFunc func(keys *delegationtype.SingleDelegationInfoReq, amounts *delegationtype.DelegationAmounts) error
+
+// IterateDelegationsForStakerAndAsset processes all operations
+// that require iterating over delegations for a specified staker and asset.
+func (k *Keeper) IterateDelegationsForStakerAndAsset(ctx sdk.Context, stakerID string, assetID string, opFunc DelegationOpFunc) error {
 	store := prefix.NewStore(ctx.KVStore(k.storeKey), delegationtype.KeyPrefixRestakerDelegationInfo)
-	amount := delegationtype.ValueField{Amount: sdkmath.NewInt(0)}
-	key := assetstype.GetJoinedStoreKey(stakerID, assetID)
-	if store.Has(key) {
-		value := store.Get(key)
-		k.cdc.MustUnmarshal(value, &amount)
-	}
+	iterator := sdk.KVStorePrefixIterator(store, delegationtype.GetDelegationStateIteratorPrefix(stakerID, assetID))
+	defer iterator.Close()
 
-	err := assetstype.UpdateAssetValue(&amount.Amount, &opAmount)
-	if err != nil {
-		return err
+	for ; iterator.Valid(); iterator.Next() {
+		var amounts delegationtype.DelegationAmounts
+		k.cdc.MustUnmarshal(iterator.Value(), &amounts)
+		keys, err := delegationtype.ParseStakerAssetIDAndOperatorAddrFromKey(iterator.Key())
+		if err != nil {
+			return err
+		}
+		err = opFunc(keys, &amounts)
+		if err != nil {
+			return err
```

Comment:
Here, if opFunc execution fails, an error is returned immediately and subsequent delegates are not processed

Reply:
The error should be returned to abort the iteration when calling this function, and the caller will handle the returned error.